File: Chapter 4 (Local search)/Basic/C++/local_beam_search.cpp
#include <bits/stdc++.h>
using namespace std;
int f(int x) { return (x - 3) * (x - 3); }
vector<int> neighbors(int x) { return {x - 1, x + 1}; }
pair<int, int> local_beam(int k, int max_steps)
{
    mt19937 rng(123);
    uniform_int_distribution<int> dist(-50, 50);
    vector<int> states(k);
    for (int i = 0; i < k; i++)
        states[i] = dist(rng);
    vector<int> values(k);
    for (int i = 0; i < k; i++)
        values[i] = f(states[i]);
    for (int step = 0; step < max_steps; step++)
    {
        vector<pair<int, int>> pv;
        for (int i = 0; i < k; i++)
            pv.push_back({values[i], states[i]});
        sort(pv.begin(), pv.end());
        int bestv = pv[0].first;
        vector<int> new_states;
        for (int i = 0; i < k; i++)
        {
            int s = pv[i].second;
            auto nbr = neighbors(s);
            for (int y : nbr)
                new_states.push_back(y);
        }
        if (new_states.empty())
            break;
        vector<pair<int, int>> cand;
        for (int s : new_states)
            cand.push_back({f(s), s});
        sort(cand.begin(), cand.end());
        states.clear();
        values.clear();
        for (int i = 0; i < k && i < (int)cand.size(); i++)
        {
            values.push_back(cand[i].first);
            states.push_back(cand[i].second);
        }
        if (values[0] >= bestv)
            break;
    }
    return {states[0], values[0]};
}
int main()
{
    auto res = local_beam(3, 100);
    cout << res.first << " " << res.second << "\n";
}

File: Chapter 4 (Local search)/Basic/Python/local_beam_search.py
import random

def local_beam_search(f,neighbors,init,k,max_steps):
    states=[init() for _ in range(k)]
    values=[f(s) for s in states]
    for _ in range(max_steps):
        pairs=list(zip(values,states))
        pairs.sort()
        bestv=pairs[0][0]
        new_states=[]
        for _,s in pairs[:k]:
            for y in neighbors(s):
                new_states.append(y)
        if not new_states:
            break
        values=[f(s) for s in new_states]
        pairs=list(zip(values,new_states))
        pairs.sort()
        states=[s for _,s in pairs[:k]]
        values=[v for v,_ in pairs[:k]]
        if values[0]>=bestv:
            break
    return states[0],values[0]

if __name__=="__main__":
    def f(x): return (x-3)*(x-3)
    def neighbors(x): return [x-1,x+1]
    init=lambda: random.randint(-50,50)
    print(local_beam_search(f,neighbors,init,3,100))